Main web3 package - 4.0.0 Alpha.1 System Tests

nikoulai opened the pull request:
## Description

#4851 

<!--
Optional if an issue is fixed:
Fixes #(issue)
-->

## Type of change

<!-- Please delete options that are not relevant. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:unit` with success.
- [ ] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.

nikoulai committed:
Config docker, Ganache and Geth

nikoulai committed:
Add some initial tests

nikoulai committed:
Add bash script for permissions

nikoulai committed:
Use local geth for integration tests

-Add commands to install and run geth for ubuntu
-Change github workflow to use local geth instead of docker

The Reason: ipc socket is not accessible outside docker

nikoulai committed:
Filter out ipc when testing ganache

nikoulai committed:
Fix ipc test

nikoulai committed:
Test workflow

nikoulai committed:
Remove verbose from jest

nikoulai committed:
Change when to init geth

nikoulai committed:
Add nohup again

nikoulai committed:
move stdout and stderr on geth-local command

nikoulai committed:
Add nohup again

nikoulai committed:
Run geth in background with screen

nikoulai committed:
Merge

nikoulai committed:
Add test using external wallet provider

nikoulai committed:
Temp solution to stop jest from hanging

nikoulai committed:
Test browser tests

nikoulai committed:
Karma config for integration tests

nikoulai committed:
Karma config for integration tests

nikoulai committed:
More (unsuccessfull) tests for karma

avkos commented on .github/Dockerfile:
```diff
@@ -1,5 +1,8 @@
-RUN geth init  --datadir data /genesis.json
+RUN geth init  --datadir data /genesis.json 
 EXPOSE 8545
+EXPOSE 8546
```
you can just remove all expose ports here. all necessary ports expose in ethereum/client-go

avkos commented on .github/workflows/build.yml:
```diff
@@ -60,10 +60,10 @@ jobs:
 
             -   if: (env.TEST_CMD == 'e2e_geth')
-                run: yarn geth:build && yarn geth && yarn test:integration
+                run: WSPORT=8546 yarn geth:install-ubuntu && yarn geth-local && yarn test:integration
```
use the same port for the websocket. 8545

avkos commented on package.json:
```diff
@@ -36,39 +36,51 @@
+		"geth:install-ubuntu": "sudo add-apt-repository -y ppa:ethereum/ethereum && sudo apt-get update && sudo apt-get install ethereum && geth init --datadir .github/geth/data `pwd`/.github/geth/genesis.json && sudo apt install screen",
 		"geth:build": "docker build ./.github -t gethnode -f .github/Dockerfile",
-		"geth": "docker run -d -p 8545:8545 gethnode --datadir data --networkid 15 --http.addr 0.0.0.0 --http.port 8545 --http",
+		"geth-local": "screen -dmS geth geth --datadir .github/geth/data --networkid 15 --http.addr 0.0.0.0 --http.port 8545 --http --ws --ws.addr 0.0.0.0 --ws.port 8546 --ws.origins '*' </dev/null &>/dev/null &",
```
use the same port for ws and http

nikoulai committed:
Batch request tests

- Expose BatchRequest class in web3 as part of Web3Context
- Add integration tests for BatchRequest

nikoulai committed:
Merge branch 'nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests' of https://github.com/ChainSafe/web3.js into nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests

nikoulai committed:
Corrent ipc path argument in geth-local

nikoulai committed:
Add wallet provider test

nikoulai committed:
Change test file filtering function so it can work on array

nikoulai committed:
Merge branch 'nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests' of https://github.com/ChainSafe/web3.js into nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests

nikoulai committed:
Filter out ws

nikoulai committed:
Rename batch request test

nikoulai committed:
Add more tests

nikoulai committed:
Geth config changes

- Use same port for ws and http
- Remove unecessary exposes port exposes
- Remove permission in docker folder

nikoulai committed:
Resolve merge

nikoulai committed:
Resolve merge

nikoulai commented:
It seems that there is a problem in docker exposing unix sockets, so I had to add scripts to install geth locally.
https://github.com/docker/for-mac/issues/483
https://github.com/ethereum/go-ethereum/issues/17907
Also, ganache doesn't support ipc, for that I had to add the filter option in jest, to ignore all tests files regarding ipc
https://github.com/trufflesuite/ganache/issues/759

nikoulai committed:
Merge branch 'nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests' of https://github.com/ChainSafe/web3.js into nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests

nikoulai committed:
Change path

nikoulai committed:
Add screen session name/fix command

jdevcs reviewed (commented):
some of e2e tests needs to be be fixed.

nazarhussain commented on package.json:
```diff
@@ -36,39 +36,51 @@
+		"geth:install-ubuntu": "sudo add-apt-repository -y ppa:ethereum/ethereum && sudo apt-get update && sudo apt-get install ethereum && geth init --datadir .github/geth/data `pwd`/.github/geth/genesis.json && sudo apt install screen",
 		"geth:build": "docker build ./.github -t gethnode -f .github/Dockerfile",
-		"geth": "docker run -d -p 8545:8545 gethnode --datadir data --networkid 15 --http.addr 0.0.0.0 --http.port 8545 --http",
+		"geth-local": "screen -DmS geth geth --datadir .github/geth/data --networkid 15 --http.addr 0.0.0.0 --http.port 8545 --http --ws --ws.addr 0.0.0.0 --ws.port 8545 --ws.origins '*' --ipcpath=./.github/geth/data/geth.ipc </dev/null &>/dev/null &",
```
Why we need to attach/reattach a session? Every CI/CD task should have an isolated scope.

nikoulai replied:
I need somehow to start geth on the background and continue with the tests https://github.com/ChainSafe/web3.js/blob/a71c534d57f5cccadd7fe9f949fa0444d6b5b709/.github/workflows/build.yml#L66

avkos replied:
I've added both starts local and background
https://github.com/ChainSafe/web3.js/pull/4895/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519
```json
"geth": "docker run -p 8545:8545 gethnode --datadir data --networkid 15 --http --http.addr 0.0.0.0 --http.port 8545 --ws --ws.addr 0.0.0.0 --ws.port 8545",
"geth-background": "docker run -d -p 8545:8545 gethnode --datadir data --networkid 15 --http --http.addr 0.0.0.0 --http.port 8545 --ws --ws.addr 0.0.0.0 --ws.port 8545",

avkos replied:
ah.you need all this of ipc provider....

nazarhussain commented on package.json:
```diff
@@ -36,39 +36,51 @@
+		"geth:install-ubuntu": "sudo add-apt-repository -y ppa:ethereum/ethereum && sudo apt-get update && sudo apt-get install ethereum && geth init --datadir .github/geth/data `pwd`/.github/geth/genesis.json && sudo apt install screen",
 		"geth:build": "docker build ./.github -t gethnode -f .github/Dockerfile",
-		"geth": "docker run -d -p 8545:8545 gethnode --datadir data --networkid 15 --http.addr 0.0.0.0 --http.port 8545 --http",
+		"geth-local": "screen -DmS geth geth --datadir .github/geth/data --networkid 15 --http.addr 0.0.0.0 --http.port 8545 --http --ws --ws.addr 0.0.0.0 --ws.port 8545 --ws.origins '*' --ipcpath=./.github/geth/data/geth.ipc </dev/null &>/dev/null &",
```
WS port should not be different than http port?

nikoulai replied:
both ganache and geth can run ws and HTTP on the same port, just for simplicity. Should I undo the change?
@avkos

avkos replied:
yes it works on the same port because it is a different protocol

nazarhussain commented on package.json:
```diff
@@ -36,39 +36,51 @@
 	},
 	"devDependencies": {
 		"@types/node": "*",
+		"assert": "^2.0.0",
```
Where this package is used?

nikoulai replied:
removed

nazarhussain commented on packages/web3-core/src/web3_context.ts:
```diff
@@ -78,6 +79,7 @@ export class Web3Context<
 	private _subscriptionManager?: Web3SubscriptionManager<API, RegisteredSubs>;
 	private _accountProvider?: Web3AccountProvider<Web3BaseWalletAccount>;
 	private _wallet?: Web3BaseWallet<Web3BaseWalletAccount>;
+	private static readonly _batchRequest = Web3BatchRequest;
```
A static but private? Can we not simply use the `Web3BatchRequest` directly?

nikoulai replied:
I changed it and exposed Web3BatchRequest directly

nazarhussain reviewed (changes requested):
There are a lot of tests cases which are commented. If these are not necessary then please remove these. If you plan to implement those later then use `it.skip` on those and remove the comments.

nikoulai committed:
Test karma conf

nikoulai committed:
Change tests to cover both jest and jasmine/workaround

nikoulai committed:
Clean package.json

nikoulai committed:
Expose Web3BatchRequest directly

nikoulai committed:
Re add ipc test

nikoulai committed:
Merge branch '4.x' into nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests

Spacesai1or commented on packages/web3/package.json:
```diff
@@ -30,16 +30,18 @@
 		"eslint-config-prettier": "^8.3.0",
 		"eslint-config-web3-base": "0.1.0",
 		"eslint-plugin-import": "^2.25.3",
+		"isomorphic-ws": "^4.0.1",
```
I'm confused why this is needed for this package, it doesn't seem to be used in these changes (assuming the reason is related to `tools/isomorphic-ws.js`)?

nikoulai replied:
You are right, I copied it from another branch and I thought it was required, but, it's not. I'll remove it. @avkos could you please verify it's not needed?

nazarhussain replied:
Yes it's not required as dependency in this package.

Spacesai1or commented on packages/web3/test/integration/external-string-providers/web3_instance_http_string.test.ts:
```diff
@@ -0,0 +1,12 @@
+	describe('Create Web3 class instance with string providers', () => {
+		it('should create instance with string of http provider', async () => {
+			const web3 = new Web3(externalHttpsStringProvider);
+			expect(web3).toBeInstanceOf(Web3);
```
```suggestion
			expect(web3).toBeInstanceOf(Web3);
			expect(web3.provider).toBe(externalHttpsStringProvider);
```

nikoulai replied:
`externalHttpsStringProvider` is just the string of the node (`localhost:8545` in our case`). I believe this won't work or do I miss something? Maybe, I have to change the name of the variable.

nazarhussain commented:
@nikoulai Please resolve the conflicts on this PR to review.

avkos commented on karma.config.js:
```diff
@@ -76,22 +108,18 @@ module.exports = function (config) {
+		exclude: ['**/web3_instance_ipc_string.test.ts'],
 		preprocessors: {
-			// Use webpack to bundle our tests files
+			'./karma-setup.js': ['webpack'],
```
I think, Should be `'./karma.setup.js': ['webpack', 'browserify'],`

nikoulai committed:
Merge branch '4.x' into nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests

nikoulai committed:
Add some headers

nikoulai committed:
Merge branch '4.x' into nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests

nikoulai committed:
WIP: Refactoring to new test infrastructure

nikoulai committed:
Merge branch '4.x' into nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests

nikoulai committed:
Refactoring to new test infrastructure

luu-alex commented on packages/web3/test/fixtures/config.ts:
```diff
@@ -0,0 +1,41 @@
+export const wsStringProvider = 'ws://localhost';
+export const ipcStringProvider = '../../../../../.github/geth/data/geth.ipc';
+export const externalHttpsStringProvider =
+	'https://mainnet.infura.io/v3/6120b2e72c304b4eadafe2bf33862ac4';
```
infura keys shouldn't be exposed

avkos replied:
So if it is committed once - it is saved in git forever. So now you have only one way. Remove and create a new one.

nazarhussain replied:
If it's in your branch, you can rewrite the commits and force-push to change the history.

nikoulai replied:
it's a node of mine

luu-alex replied:
Yes it is a node of yours, but your node is exposed on github on a huge project and may lead to vulnerabilities if someone decides to use your node. For example if a bad actor maliciously uses your node to continuosly spam requests, they may hit your infura limit and our testcases using your node will fail.
In this case we use github repo [secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets#accessing-your-secrets) so switch out your personal infura node with the github repo secret

avkos commented on packages/web3/test/fixtures/system_test_utils.ts:
```diff
@@ -0,0 +1,115 @@
+/*
+This file is part of web3.js.
```
You do not need to copy this file. Just create a symlink (like packages/web3-eth/test/fixtures/system_test_utils.ts). the fastest way is just to copy symlink from another package and paste it to your package, but you should use Finder for this

nazarhussain replied:
Just create a sym-link from root level `fixtures` directory or copy the link from some other package.

avkos commented on packages/web3/test/integration/main_web3.test.ts:
```diff
@@ -0,0 +1,249 @@
+			});
+		},
+	);
+	describeIf(getSystemTestProvider().startsWith('http'))('Web3 providers', () => {
```
looks like this test is not related to the type of provider. HTTP or WS

avkos commented on packages/web3/test/integration/main_web3.test.ts:
```diff
@@ -0,0 +1,249 @@
+			const erc20Contract = new web3.eth.Contract(erc20Abi);
+			expect(erc20Contract).toBeInstanceOf(Contract);
+		});
+		// it.skip('able to query via eth pacakge', async () => {
```
Why this test is commented on?

nazarhussain committed:
Merge branch '4.x' into nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests

nazarhussain commented on .github/geth/genesis.json:
```diff
@@ -0,0 +1,26 @@
+	"mixhash": "0x0000000000000000000000000000000000000000000000000000000000000000",
+	"parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
+	"timestamp": "0x00"
+}
```
Do we really need this file now?

nazarhussain commented on .github/setup.sh:
```diff
@@ -0,0 +1,34 @@
+}
+
+
+main "$@"
```
Same for this file? Do we really need these files now?

nazarhussain replied:
I don't see the usage of this file anywhere. Please remove this file.

nazarhussain commented on jest.config.js:
```diff
@@ -2,20 +2,19 @@ const path = require('path');
 const { lstatSync, readdirSync } = require('fs');
```
Please add license headers to this file.

nazarhussain commented on packages/web3-core/src/web3_context.ts:
```diff
@@ -235,6 +236,10 @@ export class Web3Context<
 	}
+
+	public get BatchRequest() {
+		return Web3BatchRequest.bind(null, this._requestManager as unknown as Web3RequestManager);
```
It should be a strong type `Web3RequestManager<API>`

nazarhussain commented on packages/web3/package.json:
```diff
@@ -30,16 +30,18 @@
 		"test:watch": "npm test -- --watch",
 		"test:unit": "jest --config=./test/unit/jest.config.js",
-		"test:integration": "jest --config=./test/integration/jest.config.js --runInBand --passWithNoTests"
+		"test:integration": "jest --config=./test/integration/jest.config.js --runInBand --forceExit"
```
Try if you can avoid using `--forceExit` by properly closing the socket connections. This options sometimes hides floating promises.

nazarhussain commented on packages/web3/test/fixtures/config.ts:
```diff
@@ -0,0 +1,41 @@
+		balance: '100',
+	},
+];
+export const httpStringProvider = 'http://localhost:8545';
+export const wsStringProvider = 'ws://localhost';
```
Try not to use these hardcoded values. Instead use the utility function `getSystemTestProvider()`

nazarhussain commented on packages/web3/test/fixtures/erc20.ts:
```diff
@@ -0,0 +1,238 @@
+You should have received a copy of the GNU Lesser General Public License
+along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
+*/
+export const erc20Abi = [
```
The `erc20` and `erc721` abis and other related stuff is defined in shared fixtures. Link the root level `fixtures` directory to this package as `shared_fixtures` and use these values from there.

nikoulai replied:
Currently, they are not. They exist only in `web3-eth-contract`, should I move them to `fixtures` and create symlinks in both packages symlinks? Or I miss something?

nazarhussain commented on packages/web3/test/integration/main_web3.test.ts:
```diff
@@ -0,0 +1,249 @@
+			it('should create instance with string of ws provider', async () => {
+				web3 = new Web3(clientUrl);
+				expect(web3).toBeInstanceOf(Web3);
+				await waitForSocketState(web3, 'connected');
```
Why do we need to wait for connection in this test case?

nazarhussain commented on packages/web3/test/integration/main_web3.test.ts:
```diff
@@ -0,0 +1,249 @@
+			it('should create instance with string of external ws provider', async () => {
+				web3 = new Web3(externalWssStringProvider);
+				expect(web3).toBeInstanceOf(Web3);
+				await waitForSocketState(web3, 'connected');
```
Same for here.

nazarhussain commented on tools/filter-tests.js:
```diff
@@ -0,0 +1,12 @@
+/* eslint-disable @typescript-eslint/no-unsafe-call */
+module.exports = testPaths => {
```
I think we don't this file.

jdevcs commented on packages/web3/test/integration/main_web3.test.ts:
```diff
@@ -0,0 +1,249 @@
+				expect(web3).toBeInstanceOf(Web3);
+			});
+			it('should create instance with string of external http provider', async () => {
+				const web3 = new Web3(externalHttpsStringProvider);
```
we should use `INFURA_HTTP` here as discussed like we are using in [1.x](https://github.com/ChainSafe/web3.js/blob/2c0324af2da467ee1acb72452f20000e916e4306/scripts/js/basic_usage.js#L50)

Following nodes are avalible for your tests:
` INFURA_HTTP` `INFURA_WS`  : mainnet
` INFURA_GOERLI_HTTP` `INFURA_GOERLI_WS` : goerli nw

jdevcs commented on packages/web3/test/integration/main_web3.test.ts:
```diff
@@ -0,0 +1,249 @@
+				expect(web3).toBeInstanceOf(Web3);
+			});
+			// todo fix ipc test
+			// https://ethereum.stackexchange.com/questions/52574/how-to-connect-to-ethereum-node-geth-via-ipc-from-outside-of-docker-container
```
we can skip test for instantiation with IPC provider now in alpha release, create a new issue for that.

jdevcs commented on packages/web3/test/integration/main_web3.test.ts:
```diff
@@ -0,0 +1,249 @@
+			expect(erc20Contract).toBeInstanceOf(Contract);
+		});
+		// it.skip('able to query via eth pacakge', async () => {
+		// 	const web3 = new Web3(httpStringProvider);
```
@nikoulai for remaining tests you may use similar https://github.com/ChainSafe/web3.js/pull/5015/files#diff-865c618d6dea5b91d2bc10ea4ea0491a39278e97532528ec5c33e56bc198f79bR68 via main web3 package instead of specific packages, for checking list providers given provider current providers

nikoulai committed:
Refactor deferred promises

nikoulai committed:
Revert deferred promise to it's initial implementation (not before previous commit)

nikoulai committed:
Apply changes

nikoulai committed:
Merge branch '4.x' into nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests

nikoulai committed:
Merge branch '4.x' into nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests

nikoulai committed:
fix Greeter and yarn.lock

nikoulai committed:
Add infura secrets and fix class

nikoulai committed:
Possible fix build workflow

nikoulai committed:
Fix BatchRequest function type

nikoulai committed:
Add abi encodeParameters test

nikoulai committed:
Merge branch '4.x' into nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests

nikoulai committed:
Merge branch '4.x' into nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests

nikoulai committed:
Add more provider tests

nazarhussain commented on packages/web3/test/integration/main_web3.test.ts:
```diff
@@ -0,0 +1,303 @@
+				process.env.INFURA_GOERLI_WS
+					? process.env.INFURA_GOERLI_WS.toString().includes('ws')
+					: false,
+			)('should create instance with string of external http provider', async () => {
```
```suggestion
			)('should create instance with string of external ws provider', async () => {
```

nazarhussain commented on packages/web3/test/integration/main_web3.test.ts:
```diff
@@ -0,0 +1,303 @@
+		it('providers', async () => {
+			const res = Web3.providers;
+
+			expect(res.HttpProvider).toBeDefined();
+			expect(res.WebsocketProvider).toBeDefined();
+			expect(res.IpcProvider).toBeDefined();
```
Please match these wit exact classes e.g. `WebsocketProvider`, `HttpProvider`

nazarhussain commented on packages/web3/test/integration/main_web3.test.ts:
```diff
@@ -0,0 +1,303 @@
+		});
+	});
+
+	describeIf(getSystemTestProvider().startsWith('http'))(
```
Is this necessary to only try with `http`?

nazarhussain commented on packages/web3/test/integration/main_web3.test.ts:
```diff
@@ -0,0 +1,303 @@
+					decodeParameters: expect.any(Function),
+					decodeLog: expect.any(Function),
+				}),
+			);
```
Such expectations better match with the unit tests. For integration would be better to test the user expectations from these functions.

nikoulai committed:
change module tests

nikoulai committed:
Merge branch 'nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests' of https://github.com/ChainSafe/web3.js into nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests

nikoulai committed:
Merge branch '4.x' into nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests

nazarhussain commented on packages/web3-common/src/deferred_promise.ts:
```diff
@@ -15,16 +15,9 @@ You should have received a copy of the GNU Lesser General Public License
 along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
 */
 
+/* eslint-disable max-classes-per-file */
```
No need for this comment anymore. 

```suggestion
```

nazarhussain commented on packages/web3/test/integration/main_web3.test.ts:
```diff
@@ -0,0 +1,303 @@
+		it('should execute batch requests', async () => {
+			web3 = new Web3(clientUrl);
+
+			// eslint-disable-next-line @typescript-eslint/no-floating-promises, @typescript-eslint/no-unsafe-call
```
Please don't disable `floating-promises` lint warnings. These are really evil in JS code and can have unseen side effects.

nazarhussain commented on scripts/system_tests_utils.ts:
```diff
@@ -45,7 +45,7 @@ export const getSystemTestMnemonic = (): string => getEnvVar('WEB3_SYSTEM_TEST_M
 export const getSystemTestBackend = (): string => getEnvVar('WEB3_SYSTEM_TEST_BACKEND') ?? '';
 
-const accounts: {
+export const accounts: {
```
Don't export this member, rather use `getSystemTestAccounts` helper.

nikoulai committed:
Changes from feedback

-remove lint disable for floating promises
-remove unecessary setup.sh file
-stop exporting accounts from system_utils
-use createNewAccount function for tests

jdevcs commented on packages/web3/test/integration/main_web3.test.ts:
```diff
@@ -0,0 +1,301 @@
+		});
+	});
+
+	describe('Batch Request', () => {
```
Batch Request tests are not passing, if there is any bug in batch request, could you take out this from this PR and create new issue for testing Batch Request in new issue + PR.

nikoulai committed:
Fix batch request test

nikoulai committed:
Merge branch '4.x' into nikos/4851/main-web3-package-4.0.0-alpha-1-system-tests
